sql/expression/function/distance.go

```diff
@@ -0,0 +1,185 @@
+func (d *Distance) Type() sql.Type {
+	return sql.Float64
+}
+
+func (d *Distance) String() string {
+	var args = make([]string, len(d.ChildExpressions))
+	for i, arg := range d.ChildExpressions {
+		args[i] = arg.String()
+	}
+	return fmt.Sprintf("ST_DISTANCE(%s)", strings.Join(args, ","))
+}
+
+// WithChildren implements the Expression interface.
+func (d *Distance) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	return NewDistance(children...)
+}
+
+// FindGeometryDistance finds the minimum distance between points of any geometry types a and b
+func FindGeometryDistance(a, b interface{}) float64 {
+	// Recurse until a is a Point
+	_a := sql.Point{}
+	switch a := a.(type) {
+	case sql.Point:
+		_a = a
+	case sql.Linestring:
+		res := math.MaxFloat64
+		for _, p := range a.Points {
+			res = math.Min(res, FindGeometryDistance(p, b))
+		}
+		return res
+	case sql.Polygon:
+		res := math.MaxFloat64
+		for _, l := range a.Lines {
+			res = math.Min(res, FindGeometryDistance(l, b))
+		}
+		return res
+	}
+
+	// Recurse until b is a Point
+	_b := sql.Point{}
+	switch b := b.(type) {
+	case sql.Point:
+		_b = b
+	case sql.Linestring:
+		res := math.MaxFloat64
+		for _, p := range b.Points {
+			res = math.Min(res, FindGeometryDistance(a, p))
+		}
+		return res
+	case sql.Polygon:
+		res := math.MaxFloat64
+		for _, l := range b.Lines {
+			res = math.Min(res, FindGeometryDistance(a, l))
+		}
+		return res
+	}
+
+	// Return distance between two points
+	return math.Sqrt(math.Pow(_a.X-_b.X, 2) + math.Pow(_a.Y-_b.Y, 2))
+
+}
+
+// Eval implements the sql.Expression interface.
+func (d *Distance) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	// Evaluate first geometry
+	g1, err := d.ChildExpressions[0].Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return nil when geometry is nil
+	if g1 == nil {
+		return nil, nil
+	}
+
+	// Must be a geometry type
+	switch g1.(type) {
+	case sql.Point, sql.Linestring, sql.Polygon:
+	default:
+		return nil, sql.ErrInvalidGISData.New(d.FunctionName())
+	}
+
+	// Evaluate second geometry
+	g2, err := d.ChildExpressions[1].Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return nil when geometry is nil
+	if g2 == nil {
+		return nil, nil
+	}
+
+	// Must be a geometry type
+	switch g2.(type) {
+	case sql.Point, sql.Linestring, sql.Polygon:
+	default:
+		return nil, sql.ErrInvalidGISData.New(d.FunctionName())
+	}
+
+	// Find distance between geometry
+	dist := FindGeometryDistance(g1, g2)
+
+	// If no unit argument, return
+	if len(d.ChildExpressions) == 2 {
+		return dist, nil
+	}
+
+	// Evaluate units
+	unit, err := d.ChildExpressions[2].Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// if unit is null, return null
+	if unit == nil {
+		return nil, nil
+	}
+
+	// unit must be string
+	_unit, err := sql.LongText.Convert(unit)
+	if err != nil {
+		return nil, err
+	}
+
+	// Must be a valid unit
+	switch _unit.(string) {
```

Comment:
Given that there's no happy path here, I would just move this check to the constructor.